kuzzleio/kuzzle
# Add support for compressed HTTP requests (gzip, deflate)

:warning: HTTP Proxy functional tests will fail until https://github.com/kuzzleio/kuzzle-proxy/pull/100 is merged (and docker images are updated)

## What does this PR do?

* Add support for incoming compressed HTTP requests (accepted algorithms: gzip, deflate). As per the RFC, the `content-encoding` headers contains the *list* of compression algorithms used, in the order used to compress the data (a request may be compressed with multiple layers of gzip and/or deflate)
* Add support for outgoing compressed HTTP requests. This is controlled by the requesting client, using the `accept-encoding` header. Accepted algorithms: gzip, deflate. Because deflate is often badly implemented by clients, gzip is preferred when given a choice 
* The HTTP headers sent by default by Kuzzle now contain an `accept-encoding` headers
* Two new configuration parameters have been added allowing administrators to control this feature:
  * `server.protocols.http.allowCompression` (boolean, default: `true`): used to enable/disable the feature for *incoming* requests
  * `server.protocols.http.maxEncodingLayers` (integer, default: `3`): limits the number of compression layers, to prevent cpu/memory bombs

### How should this be manually tested?

New HTTP-only functional tests have been added, automatically testing the feature.
If you want to check manually, use your favorite HTTP client and play with the `content-encoding` and `accept-encoding` headers

### Other changes

* Reading HTTP incoming messages is now properly done, using stream pipes, instead of emulating that feature with events
* Promoted the `entryPoint` property in the protocol implementations in the abstract Protocol class, DRYifying it

### Boyscout

* Add more extensive HTTP options documentation in the `.kuzzlerc.sample` file
* Add missing headers in the `Access-Control-Allow-Headers` list (fix #1135)
